app/modules/auth/reset_password/reset_password.controller.ts

```diff
@@ -0,0 +1,112 @@
+import { oneLine } from 'common-tags';
+import { IRequest, IResponse } from '../../../interfaces/utils/IServer';
+import { EmailTypes, UserStatuses } from '../../../interfaces/enums';
+import { sendEmail } from '../../../../utils/email_service';
+import { generatePasswordResetLink } from '../../../../utils/helpers';
+
+export default class ResetPasswordController implements IController {
+  public async post(req: IRequest, res: IResponse, next: restify.Next) {
+    try {
+      const { email } = req.body;
+
+      const user = <IUser> await User.findOne({ where: <IUser>{ email } });
+
+      if (!user) {
+        return res.send(new NotFoundError(`User doesn't exist in our db`));
+      }
+
+      const { link, token } = generatePasswordResetLink();
+
+      await User.update(<IUser> {
+        reset_password_token: hashSync(token, 12),
+        reset_password_attempts: 0
+      }, { where: { id: user.id } });
+
+      await sendEmail(user.email, user.name, EmailTypes.RESET_PASSWORD, { link, token });
+
+      return res.send({
+        success: true,
+        message: oneLine`
+          Password reset email is successfully sent. Please check your email and follow
+          the instructions to reset your password.
+        `
+      });
+    } catch (e) {
+      req.log.error(e);
+      return res.send(new InternalServerError(e));
+    }
+    return next();
+  }
+
+  public async put(req: IRequest, res: IResponse, next: restify.Next) {
+    try {
+      const { email, new_password, reset_token } = req.body;
+      const user = <IUser> await User.findOne({ where: <IUser>{ email } });
+
+      if (!user) {
+        return res.send(new NotFoundError(`User doesn't exist in our db`));
+      }
+
+      if (!user.reset_password_token) {
+        return res.send({
+          error: true,
+          message: oneLine`
+            This password reset token was either used or has expired. Please try
+            generating new password reset token and try again.
+          `
+        });
+      }
+
+      if (user.isMaxPasswordResetAttemptReached()) {
+        // Change the account status to suspened, when the password is successfully reset
+        // it'll be set to active
+        await User.update(<IUser> {
+          status: UserStatuses.SUSPENDED
+        }, { where: { id: user.id } });
+
+        return res.send({
+          error: true,
+          message: oneLine`
+            Your account has reached max invalid password reset attempts. Please try
+            requesting new password reset link again.
+          `
+        });
+      }
+
+      if (!compareSync(reset_token, user.reset_password_token)) {
+        // Update the invalid password reset attempts
+        await User.update(<IUser> {
+          reset_password_attempts: user.reset_password_attempts + 1
+        }, { where: { id: user.id } });
+
+        return res.send({
+          error: true,
+          message: 'Given Password reset token is incorrect'
```

Comment:
A check for reset_password_attempts should be added here. In case this equals 5, the user's status should be set to `suspended` immediately so other users can't perform all functions with regards to this user (account transfer, user information, wallet sharing, etc.).